packages/ethereum-group-pcd/src/EthereumGroupPCD.ts

```diff
@@ -0,0 +1,334 @@
+    },
+  });
+
+  return new EthereumGroupPCD(
+    uuid(),
+    {
+      publicInput: publicInput,
+      groupType:
+        args.groupType.value === "address"
+          ? GroupType.ADDRESS
+          : GroupType.PUBLICKEY,
+    },
+    {
+      signatureProof: await SemaphoreSignaturePCDPackage.serialize(
+        semaphoreSignature
+      ),
+      ethereumGroupProof: Buffer.from(proof).toString("hex"),
+    }
+  );
+}
+
+export async function verify(pcd: EthereumGroupPCD): Promise<boolean> {
+  const semaphoreSignature = await SemaphoreSignaturePCDPackage.deserialize(
+    pcd.proof.signatureProof.pcd
+  );
+  const signatureProofValid = await SemaphoreSignaturePCDPackage.verify(
+    semaphoreSignature
+  );
+
+  // the semaphore signature of the group membership proof must be valid
+  if (!signatureProofValid) {
+    return false;
+  }
+
+  // The message signed by the semaphore signature must be the same as the
+  // serialized ethereum group proof
+  if (semaphoreSignature.claim.signedMessage !== pcd.proof.ethereumGroupProof) {
+    return false;
+  }
+
+  const deserializedSignatureProof =
+    await SemaphoreSignaturePCDPackage.deserialize(
+      pcd.proof.signatureProof.pcd
+    );
+
+  const deserializedIdentity =
+    deserializedSignatureProof.claim.identityCommitment;
+  const message = deserializedIdentity;
+  const msgHash = ethers.utils.hashMessage(message);
+
+  const publicInputMsgHash =
+    "0x" + pcd.claim.publicInput.msgHash.toString("hex");
+
+  // The string that the ethereum group proof was signed over must the same
+  // as the semaphore identity commitment. (Their hashes are equal)
+  if (msgHash !== publicInputMsgHash) {
+    return false;
+  }
+
+  const verifier =
+    pcd.claim.groupType === GroupType.ADDRESS ? addrVerifier : pubkeyVerifier;
+  const groupMembershipValid = await verifier.verify(
+    Buffer.from(pcd.proof.ethereumGroupProof, "hex"),
+    pcd.claim.publicInput.serialize()
+  );
+
+  if (!groupMembershipValid) {
+    return false;
+  }
+
+  return true;
+}
+
+export async function serialize(
+  pcd: EthereumGroupPCD
+): Promise<SerializedPCD<EthereumGroupPCD>> {
+  return {
+    type: EthereumGroupPCDTypeName,
+    pcd: JSONBig({ useNativeBigInt: true }).stringify(pcd),
+  } as SerializedPCD<EthereumGroupPCD>;
+}
+
+export async function deserialize(
+  serialized: string
+): Promise<EthereumGroupPCD> {
+  const parsed = JSONBig({ useNativeBigInt: true }).parse(serialized);
+  const publicInput = new PublicInput(
+    parsed.claim.publicInput.r,
+    parsed.claim.publicInput.rV,
+    Buffer.from(parsed.claim.publicInput.msgHash),
+    new CircuitPubInput(
+      parsed.claim.publicInput.circuitPubInput.merkleRoot,
+      parsed.claim.publicInput.circuitPubInput.Tx,
+      parsed.claim.publicInput.circuitPubInput.Ty,
+      parsed.claim.publicInput.circuitPubInput.Ux,
+      parsed.claim.publicInput.circuitPubInput.Uy
+    )
+  );
+  parsed.claim.publicInput = publicInput;
+  return parsed;
```

Comment:
One issue with this `parse` function is that it's not invoking the constructor of `EthereumGroupPCD`, but creating an object that conforms to its type signature. Not a huge deal right now but feels kind of wrong. Checking the way I do this in other packages and it seems that this is a pervasive error. Here's an issue to fix at some point later. https://github.com/proofcarryingdata/zupass/issues/308